From ae5086faf6cca1f7719d34db718d3a4bfc9de2a2 Mon Sep 17 00:00:00 2001
From: Marek Vasut <marex@denx.de>
Date: Tue, 19 Nov 2019 00:45:25 +0100
Subject: [PATCH 3/4] etnaviv: Check for icache availability if required

On lower-end GPUs without the icache, the maximum shader instruction count
is the hard limit. In case the shader has more instructions, indicate that
the shader compilation failed.

Signed-off-by: Marek Vasut <marex@denx.de>
---
 src/gallium/drivers/etnaviv/etnaviv_compiler_nir.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/gallium/drivers/etnaviv/etnaviv_compiler_nir.c b/src/gallium/drivers/etnaviv/etnaviv_compiler_nir.c
index 8a71f62a07d..1210c2f1fdb 100644
--- a/src/gallium/drivers/etnaviv/etnaviv_compiler_nir.c
+++ b/src/gallium/drivers/etnaviv/etnaviv_compiler_nir.c
@@ -814,6 +814,9 @@ etna_compile_shader_nir(struct etna_shader_variant *v)
    v->code = code;
    v->needs_icache = c->inst_ptr > specs->max_instructions;
 
+   if (!c->specs->has_icache && v->needs_icache)
+      return false;
+
    copy_uniform_state_to_shader(v, c->consts, num_consts);
 
    if (s->info.stage == MESA_SHADER_FRAGMENT) {
-- 
2.24.1

